### ISSUE TITLE
ignore specific metadata_entries

### ISSUE THREAD

**Issue by bogi0704** [vmware/terraform-provider-vcd#959]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

I am migrating our use of the now deprecated `metadata` to the new `metadata_entry` block. I have the need to ignore specific `metadata_entries`, because they are managed by other systems rather than terraform. There is for example the `veeam-plugin` for VCD, which itself creates a `metadata_entry` of `veeam-endpoint` -> `https://<URL>` in the `vcd_org` object. The same is the case for `object-storage-extensions`. We also have our own processes which write their metadata to the `vcd_org` or `vcd_org_vdc` objects.

I was able to work around this until now, because of the way the old `metadata`-entries were working. The following workaround was working:

```terraform
lifecycle {
    ignore_changes = [
      metadata["ose-storage-id"], metadata["veeam-endpoint"]
    ]
  }
```

Using the new `metadata_entry`, there is no such way of ignoring specific entries. All the ways I have tried have failed. This is because I am now required to evaluate the ignore_changes values in a more dynamic way, which currently isn't possible. 

### New or Affected Resource(s)

All resources supporting `metadata_entry`. 

### Example behaviour

An example behaviour after migrating to `metadata_entry` and performing a plan:

```hcl
# module.*.vcd_org.org will be updated in-place
  ~ resource "vcd_org" "org" {
        id                              = "urn:vcloud:org:65ce7ee1-a2e0-49d0-b21f-2d86121d6f9a"
        - metadata_entry {
          - is_system   = true -> null
          - key         = "veeam-endpoint" -> null
          - type        = "MetadataStringValue" -> null
          - user_access = "READONLY" -> null
          - value       = "\"https://backup.<URL>:9443\"" -> null
        }
```

### References

There is a rather old request to improve this on the terraform core, but it looks like this will not be happening: https://github.com/hashicorp/terraform/issues/3116 

The AWS provider has implemented a workaround for this, see https://registry.terraform.io/providers/hashicorp/aws/latest/docs/guides/resource-tagging#ignoring-changes-in-all-resources . This would also allow to ignore specific prefixes, which would be even more flexible than the old way I was doing it.

I will not be able to migrate to `metadata_entry`, until this is resolved. I would brick our tenant's ability to use the VCD extensions for veeam, object storage and so on.

**Comment by jpbuecken** [comment]
Hello,
I had the same use case today. Cloud Directory Availability add the metadata entry:

```
      - metadata_entry {
          - is_system   = true -> null
          - key         = "vmware.cloud.director.availability.protected" -> null
          - type        = "MetadataStringValue" -> null
          - user_access = "PRIVATE" -> null
          - value       = "PROTECTED" -> null
        }
```

Now I want to ignore exactly this combination of a metadata_entry. But I still want to update other combination of those values, to set further metadata tags

**Comment by adambarreiro** [comment]
Hi @bogi0704, @jpbuecken,
Thanks for the proposal! We will evaluate it with the team.

Best regards.

**Comment by jpbuecken** [comment]
Hello @adambarreiro , 
not sure about your design ideas, but we have a proposal.

The NSXT provider considers provider wide ignores, but a hardcoded list in code of the provider is limited and insufficient from the beginning.
https://github.com/vmware/terraform-provider-nsxt/issues/820#issuecomment-1340159629

We already have own services and third party tools that uses metadata that we need to ignore.
And VMware may change usage of metadata as well, and the provider needs to keep up to this state.

Instead, our proposal is a metadata_entry_ignore as counterpart to metadata_entry.

So exactly this "map" can be ignored:
```
  metadata_entry_ignore" {
      key         = "vmware.cloud.director.availability.protected"
      value       = "PROTECTED"
      type        = "MetadataStringValue"
      user_access = "PRIVATE"
      is_system   = true
    }
```
If one of the key-value pairs of the map differ, the metadata entry is not ignored anymore.

**Comment by bogi0704** [comment]
I think that specifying a key and a value to ignore specifically would not fit our use-cases completely. I would like to be able to ignore specific and also wildcarded keys without regard to the value:

```
  metadata_entry_ignore" {
      key         = "vmware.cloud.director.*"
      type        = "MetadataStringValue"
      user_access = "PRIVATE"
      is_system   = true
    }
```

This should ignore all key-value pairs, if the prefix of the key fits "vmware.cloud.director.*".

More specific ignore entries should also be possible, but they should't be the only way. I need it to be more flexible.

**Comment by jpbuecken** [comment]
@bogi0704 : Wildcard support ist a good idea, to understand your hcl: All metadata_entries that matches key, type, user_access, is_system as specified will be ignored, independend of the value of `value`. 

So assume we have wildcard support for `value`, this would be the same:

```
 metadata_entry_ignore" {
      key         = "vmware.cloud.director.*"
      type        = "MetadataStringValue"
      user_access = "PRIVATE"
      is_system   = true
    }

 metadata_entry_ignore" {
      key         = "vmware.cloud.director.*"
     value = "*"
      type        = "MetadataStringValue"
      user_access = "PRIVATE"
      is_system   = true
    }
```

**Comment by dataclouder** [comment]
While wildcards support could be implemented, I suggest that –if this feature is developed– we use regular expressions instead, for consistency with other resource that use a similar way of searching.

**Comment by adambarreiro** [comment]
Hello,

We just merged #1057.

You can find the new Provider argument `ignore_metadata_changes` [here](https://github.com/vmware/terraform-provider-vcd/blob/main/website/docs/index.html.markdown), if you wish to provide your feedback, you can clone the repository and run `make install` to install the non-yet-released version of the provider with this feature.

Best regards.
